doc/langdef.md

```diff
@@ -116,44 +126,185 @@ In general it is a bad idea for those defining contexts or extensions to use
+    absolute or relative name in the protocol buffer package namespace. Package
+    names must be followed by a message type or enum constant.
+*   Protocol buffer message types and enum constants: following an optional
+    protocol buffer package name, a simple or qualified name could refer to a
+    message type or enum constant in the package's namespace.
+
+Resolution works as follows. If `a.b` is a name to be resolved in the context of
+a protobuf declaration with scope `A.B`, then resolution is attempted, in order,
+as `A.B.a.b`, `A.a.b`, and finally `a.b`. To override this behavior, one can use
+`.a.b`; this name will only be attempted to be resolved in the root scope, i.e.
+as `a.b`.
+
+If name qualification is mixed with field selection, the longest prefix of the
+name which resolves in the current lexical scope is used. For example, if
+`a.b.c` resolves to a message declaration, and `a.b` does so as well with `c` a
+possible field selection, then `a.b.c` takes priority over the interpretation
+`(a.b).c`. Explicit parentheses can be used to force the field selection
+interpretation.
+
 ## Values
 
 Values in CEL represent any of the following:
 
-1.  *Simple Values*, as they come from protocol buffers: booleans, integers,
-    floating-point numbers, strings, byte strings, and messages. Numbers are
-    normalized to the highest precision supported by protobuf, i.e. int32 as
-    int64, uint32 as uint64, and float as double. Enum values are represented as
-    integers. In addition, the well-known types `Duration` and `Timestamp` are
-    not treated as messages, but as primitive values of the given type.
-2.  *Null Value*, a special value to represent an absent value, for example, the
-    default value of messages.
-3.  *Aggregate Values*, as they are implied by `repeated` and `map` in protobuf
-    messages: lists of the above values, and maps of the above values, with the
-    restriction on allowed map keys implied by protobuf.
-4.  *Type Values*. See [Types](#types) below.
-5.  *Function Values*. See [Functions](#functions) below.
+Type          | Description
+------------- | ---------------------------------------------------------------
+`int`         | 64-bit signed integers
+`uint`        | 64-bit unsigned integers
+`double`      | 64-bit IEEE floating-point numbers
+`bool`        | Booleans (`true` or `false`)
+`string`      | Strings of UTF-8 code points
+`bytes`       | Byte sequences
+`list`        | Lists of values
+`map`         | Associative arrays with `int`, `uint`, `bool`, or `string` keys
+`null_type`   | The value `null`
+message names | Protocol buffer messages
+`type`        | Values representing the types in the first column
+
+### Numeric Values
+
+CEL supports only 64-bit integers and 64-bit IEEE double-precision
+floating-point. We only support positive, decimal integer literals; negative
+integers are produced by the unary negation operator. Note that the integer 7 as
```

Comment:
While this is true that negatives are produced via unary negation, 
I think this is a bug rather than a specific design of the language.

Reply:
If you'd write the bug, I'd be happy to fix it.  In the meantime, I'm describing the language as-is.  It's not just an implementation issue, but it reflects the actual syntax of the grammar, so I don't think we can ignore it.  If we change the grammar to have negative constants later, it will be compatible with current programs.